fix(lsp): ensure uri_from_fname always returns valid URI

clason opened the pull request:
Problem: if fname is a relative path, the returned URI is invalid (only absolute paths are allowed)
Solution: call `vim.fn.fnamemodify(fname, ':p') to force absolute path

Adapt tests so Windows file paths are only tested on Windows.

clason commented:
Hmmm... that breaks the Windows filepath tests, since the fake filenames are now treated as local files on the Unix runners.

clason commented on runtime/lua/vim/uri.lua:
```diff
@@ -59,8 +59,12 @@ end
+---@param force_valid boolean Force valid absolute paths
 ---@return string URI
-local function uri_from_fname(path)
+local function uri_from_fname(path, force_valid)
```
This seemed the safest option as there may be (and for the tests definitely are) reasons _not_ to force it.

I can flip the default, of course, to make it opt-out, but then we risk breaking downstream.

clason replied:
We can also remove the tests if we'd rather hard-code this?

justinmk replied:
Maybe the tests could check `iswin()` ?

clason replied:
you mean only check Windows path conversion on Windows (and vice versa)?

justinmk replied:
yes. Don't see a use-case for the existing behavior.

clason committed:
fix(lsp): ensure valid URI from fname

Problem: if fname is a relative path, the returned URI is invalid
         (only absolute paths are allowed)
Solution: call `vim.fn.fnamemodify(fname, ':p') to force absolute path

This means we can no longer test Windows file paths on Unix and vice
versa.

clason commented:
Breaks the `lsp_spec` tests as well (presumably due to the fake file names) ...

clason commented:
Yeah, that's not gonna work. Leaving this open for a bit in case anybody has any idea how to teach the fake server to handle valid URIs, otherwise not much we can do.

**PSA:** Make sure you always feed absolute paths to anything LSP-related as otherwise the conversion to URI will not generate valid URIs.

clason commented:
Alternative, less intrusive approach:

1. Add a note to the `uri_from_fname` docstring about relative paths not being validated and possibly leading to invalid URIs.
2. Change `:h` to `:p:h` in
https://github.com/neovim/neovim/blob/0b72e23bf13ade1e76d11eae6990014098b3928d/runtime/lua/vim/fs.lua#L43
and note that you always get an absolute path. (One could argue that this would be the more robust and expected behavior anyway?)

@gpanders ?

gpanders commented:
> 2. Change `:h` to `:p:h` in
>     https://github.com/neovim/neovim/blob/0b72e23bf13ade1e76d11eae6990014098b3928d/runtime/lua/vim/fs.lua#L43
>     
>    and note that you always get an absolute path. (One could argue that this would be the more robust and expected behavior anyway?)

`:p:h` in `fnamemodify` has unexpected behavior (e.g. `fnamemodify('/Users/greg/src', ':p:h')` returns `/Users/greg/src`, *not* `/Users/greg`). I also think the `dirname` function should be as simple/dumb as possible, given some path just strip off everything after the final `/` (which is what `:h` does). For these reasons I'm not in favor of using `:p:h` here.

clason commented:
That sounds like a bug of `fnamemodify`...

But OK, not going to debate this.

gpanders commented:
>That sounds like a bug of fnamemodify...

The `:p` converts` /Users/greg/src` to `/Users/greg/src/` and the `:h` strips off the trailing `/` to change it back to `/Users/greg/src`. So not a bug per se, just tricky/confusing/non-intuitive (maybe bad) semantics.

clason commented:
Yeah, it's even documented. Weird.
